cmd/otel-allocator/allocation/allocator.go

```diff
@@ -0,0 +1,138 @@
+package allocation
+
+import (
+	"log"
+	"sync"
+)
+
+/*
+	Load balancer will serve on an HTTP server exposing /jobs/<job_id>/targets <- these are configured using least connection
+	Load balancer will need information about the collectors in order to set the URLs
+	Keep a Map of what each collector currently holds and update it based on new scrape target updates
+*/
+
+// Create a struct that holds collector - and jobs for that collector
+// This struct will be parsed into endpoint with collector and jobs info
+
+type collector struct {
+	Name       string
+	NumTargets int
+}
+
+// Allocator makes decisions to distribute work among
+// a number of OpenTelemetry collectors based on the number of targets.
+// Users need to call SetTargets when they have new targets in their
+// clusters and call Reshard to process the new targets and reshard.
+type Allocator struct {
+	m sync.Mutex
+
+	targetsWaiting map[string]TargetItem // temp buffer to keep targets that are waiting to be processed
+
+	collectors map[string]*collector // all current collectors
+
+	targetItems map[string]*TargetItem
+}
+
+// findNextCollector finds the next collector with less number of targets.
+func (allocator *Allocator) findNextCollector() *collector {
+
+	col := &collector{}
+	for _, v := range allocator.collectors {
+		// If the initial collector is empty, set the initial collector to the first element of map
+		if col.Name == "" {
+			col = v
```

Comment:
```suggestion
	var col *collector
	for _, v := range allocator.collectors {
		// If the initial collector is empty, set the initial collector to the first element of map
		if col == nil {
			col = v
```

Can we avoid the allocation of the initial `collector`?  Perhaps conditionally create an empty `collector` at return if the concern is about avoiding nil pointer checks everywhere the result is used.

Reply:
Yeah i agree. The initial purpose of this is to have an element to refer to as the 'min'